### ISSUE TITLE
metal 1.0: API change: don't use sparse enums for CB indices (will improve fd2 perf)

### ISSUE THREAD

**Issue by pgkeller** [tenstorrent/tt-metal#7493]
Current API requires CBs to be specified by an enum whose entries may be sparsely populated within the 0..31 range
Firmware and dispatch could benefit from using densely packed indices

This is to change the API to use indices from 0..31 explicitly (or some other allocation mechanism where indices are returned?) to pack the indices as much as possible to reduce transfer and initialization overhead

**Comment by eyonland** [comment]
@rdjogoTT, as mentioned on a comment...

> The LLKs can read from/push into any CB id, because it just translates into L1 addresses. The bigger issue is we have > data_format.cpp , that does data format checks/validation, and also translates dataformats into HW specific formats, and that file makes assumptions of where the inputs ids and the output ids are.

From @davorchap,  I see these comments...

> I think “the dataflow CBs” don’t have any data format checks
> So probably not safe to use with Tensix
> @rdjogoTT would good to see if we can relax , generalize data format cpp , to allow more flexible usage of 32 CBs
> This current setup dates back all the way to spatial1

What is needed to get this unblocked?

**Comment by rdjogoTT** [comment]
To my knowledge the changes that are needed should be limited to the JIT build files, since LLKs shouldn't depend on the structure we are referring to here (32 CBs with 8 input, 8 output, ...). LLKs just need the unpack/pack src and dst format arrays to be generated with the correct src/dest format pairs.

**Comment by pgkeller** [comment]
Fixing this and fixing the CB API will save 200-400 cycles in the common case in dispatch init code, so is fairly important.  I'm not familiar w/ the changes needed in data_format.cpp.  Is this something that could easily be picked up by someone new or should we have someone from LLKs who is familiar w/ it take it on?  If the latter, any ideas when this can get scheduled?  Thanks.

**Comment by rdjogoTT** [comment]
Actually, since the goal here is just to get rid of the sparse CB enum structure it should be ok for someone new to make the changes. They can just leave the dataformat checks as are and someone from LLKs can clean those up in a separate issue to make the CBs more general and get rid of the old restrictions from old stacks.

**Comment by ttmtrajkovic** [comment]
@pgkeller , @rdjogoTT,

Current LLK code relies on CB indices to be constants in order to get compiler optimizations and faster and smaller code. This new change should ensure that "new" indices for CBs that are used are also constexpr, probably generated by JIT since that information should be known at compile time.

If I understood this correctly, we should have an additional data structure of CB indices in use and some data showing which ones are inputs and which are outputs? Bottom line, all lookups into data_format, tile size structures that LLKs do must still be compile-time optimizable.

Milos

**Comment by pgkeller** [comment]
Hmm, that's an interesting point.  The values will be known at kernel compile time but not host compile time, so for now won't be a problem.  However, as we move to offline build, using runtime indices will be a problem

**Comment by ttmtrajkovic** [comment]
I am not sure how we would do an offline build, but switching to runtime indices may have impact to perf and code size. 
Let's evaluate that before pulling the trigger on offline build

**Comment by pgkeller** [comment]
Offline build is a requirement for a number of customers, but I think we can manage it w/ the API requesting specific indices and the offline build passing those in as compile time args.  This will take some thought.
